### ISSUE TITLE
Istio sidecar not allowing connection to database

### ISSUE THREAD

**Issue by pepjo** [istio/istio#9157]
**Describe the bug**
Trying to connect to a Postgres database running inside Istio and kubernetes from a service inside istio, both with sidecars. Instead of allowing the connection the sidecar logs show:

```
[15][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:240] gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers
```

**Expected behavior**
They should be able to stablish a connection.

**Steps to reproduce the bug**
- Deploy a postgres database + service
- Deploy a service that connects to the database

**Version**
ISTIO:
Version: 1.0.2
GitRevision: d639408fded355fb906ef2a1f9e8ffddc24c3d64
User: root@66ce69d4a51e
Hub: gcr.io/istio-release
GolangVersion: go1.10.1
BuildStatus: Clean

KUBERNETES: (v1.11)
Client Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.0-rc.1", GitCommit:"3e4aee86dfaf933f03e052859c0a1f52704d4fef", GitTreeState:"clean", BuildDate:"2018-09-18T21:08:06Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-09T17:53:03Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

**Installation**
Using helm with gateways.enabled=false & tracing.enabled=true

**Environment**
Kubernetes deployed in DigitalOcean

**NOTES**
 - Only works as expected if I deactivate the sidecar using the `sidecar.istio.io/inject: "false"` annotation on BOTH deployments work
 - I know I have to store passwords as secrets for production

**Deployment files**
Postgres database:
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: postgres-config
  namespace: app
  labels:
    app: postgres
data:
  POSTGRES_DB: postgres
  POSTGRES_USER: postgres
  POSTGRES_PASSWORD: mysecretpassword
---
kind: PersistentVolume
apiVersion: v1
metadata:
  name: postgres-pv-volume
  namespace: app
  labels:
    type: local
    app: postgres
spec:
  storageClassName: manual
  capacity:
    storage: 500Mi
  accessModes:
  - ReadWriteMany
  hostPath:
    path: "/mnt/data"
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: postgres-pv-claim
  namespace: app
  labels:
    app: postgres
spec:
  storageClassName: manual
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 500Mi
---
apiVersion: v1
kind: Service
metadata:
  name: postgres
  namespace: app
  labels:
    app: postgres
spec:
  ports:
  - port: 5432
    name: db
    # protocol: TCP
  selector:
    app: postgres
  # clusterIP: None
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: postgres-deployment
  namespace: app
  labels:
    app: postgres
spec:
  replicas: 1
  selector:
    matchLabels:
      app: postgres
  template:
    metadata:
      # annotations:
      #   sidecar.istio.io/inject: "false"
      labels:
        app: postgres
    spec:
      containers:
      - name: postgres
        image: postgres:10.4
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 5432
          name: db
          # protocol: TCP
        livenessProbe:
          exec:
            command:
            - pg_isready
            - -U
            - postgres
          initialDelaySeconds: 30
          timeoutSeconds: 3
        readinessProbe:
          exec:
            command:
            - pg_isready
            - -U
            - postgres
          initialDelaySeconds: 30
          timeoutSeconds: 3
        envFrom:
        - configMapRef:
            name: postgres-config
        volumeMounts:
        - mountPath: /var/lib/postgresql/data
          name: postgredb
      volumes:
      - name: postgredb
        persistentVolumeClaim:
          claimName: postgres-pv-claim
```

Service that connects to the databse:
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: pgservice-config
  namespace: app
  labels:
    app: pgservice
data:
  POSTGRESQL_CONNECTION: jdbc:postgresql://postgres:5432/postgres
  POSTGRESQL_USER: postgres
  POSTGRESQL_PASSWORD: mysecretpassword
---
apiVersion: v1
kind: Service
metadata:
  name: pgservice
  namespace: app
  labels:
    app: pgservice
spec:
  type: ClusterIP
  selector:
    app: pgservice
  ports:
  - port: 8080
    name: http
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: pgservice-deployment
  namespace: app
  labels:
    app: pgservice
spec:
  replicas: 1
  selector:
    matchLabels:
      app: pgservice
  template:
    metadata:
      # annotations:
      #   sidecar.istio.io/inject: "false"
      labels:
        app: pgservice
    spec:
      initContainers:
      - name: check-db-ready
        image: postgres:10.4
        command: ['sh', '-c',
                  'until pg_isready -h postgres -p 5432;
                do echo waiting for database; sleep 2; done;']
      - name: exec-db-migration
        image: registry.gitlab.com/registry/pgmigration:latest
        imagePullPolicy: Always
        envFrom:
        - configMapRef:
            name: pgmigration-config
      containers:
      - name: pgservice
        image: registry.gitlab.com/registry/pgservice:latest
        imagePullPolicy: Always
        envFrom:
        - configMapRef:
            name: pgservice-config
        ports:
        - containerPort: 8080
          name: http
      imagePullSecrets:
      - name: regcred
```

**Comment by geauxvirtual** [comment]
I've got a PostgreSQL database and a service working with both the postgresql pod and the service deployed with sidecars and mutual-tls enabled between the two sidecars. In your service definition for PostgreSQL, you have ClusterIP commented out. What IP does "postgresql" resolve to when you try to connect to it from the service? If it's a pod IP and not a cluster IP, it's my understanding Istio will block this traffic. The service would need to talk to PostgreSQL over a ClusterIP.

**Comment by pepjo** [comment]
Hi @geauxvirtual ,

I commented this line to show that I already tried that option. If I am not mistaken, if I don't include the type of service kubernetes defaults to CluterIP anyway.

Just to make sure, I un-commented the line and tried again and I still see the message error in the istio sidecar:
```
[2018-10-11 12:58:41.173][15][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:240] gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers
```

I removed that comment following [this conversation in google groups](https://groups.google.com/d/msg/istio-users/-cNoLrE5904/fxCro5wSBgAJ). They propose an example at [this link](https://github.com/dockersamples/k8s-wordsmith-demo/blob/master/kube-deployment.yml). As you see, the example does not have the ClusterIP line, that is why I tried removing the ClusterIP line.

Do you have any special policies in place to make this work? What version of Istio are you running?

**Comment by geauxvirtual** [comment]
Using Istio 1.0.1. I have  a policy to enable Mutual TLS on port 5432 for the postgresql service, but I didn't have any issues connecting before enabling Mutual TLS either.

Can you exec into the istio-proxy container injected into your service and see what IP nslookup is returning for the `postgres` service you are trying to connect to?

**Comment by konokimo** [comment]
Having the same issue. When the sidecar is injected successfully into my pod, the pod is unable to reach my Cassandra DB Instance.

**Comment by herasoo** [comment]
I have same experience. DB pod is up with proxy, and SERVICE pod is loading with proxy. The problems is that the SERVICE container not proxy checks or makes DB connections during container's start-up.
So I think this situation happened before SERVICE proxy container completely started up.
In Istio, Service-Mesh, all network is possible through completely started up proxies.

Any ideas? for modifying an order to load containers in a pod and etc?

**Comment by daudn** [comment]
Did anybody find a fix for this? I cannot connect to my Datastore database while using GKE with Istio.

**Comment by herasoo** [comment]
In this case, I recommend to add below object.

apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: postgres
  namespace: app
spec:
  host: postgres
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL

**Comment by daudn** [comment]
@herasoo this is unindented, correct?

**Comment by daudn** [comment]
This is my k8's yaml configuration for "my-app" that requires to connect to Datastore on port (5432) and RabbitMQ on port (5672)
`apiVersion: v1
kind: Service
metadata:
  name: myapp-service
  labels:
    app: myapp-service
spec:
  ports:
  - name: rabbitmq
    protocol: TCP
    port: 5672
  - name: datastore
    protocol: TCP
    port: 5432
    targetPort: 5432
  type: NodePort
  selector:
    app: pbbarcode-pod
---
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
spec:
  containers:
    - name: myapp
      image: eu.gcr.io/<projectId>/myapp:latest
      imagePullPolicy: Always
      ports:
        - containerPort: 5672
        - containerPort: 5432
      env:
        - name: MY_NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
      lifecycle:
        preStop:
          exec:
            command: ["python3", "graceful_termination.py"]
      volumeMounts:
      # Name of root folder where filestore is mounted
      - mountPath: Filestore/
        name: mypvc
  nodeSelector:
    Focus: myapp
  volumes:
  - name: mypvc
    persistentVolumeClaim:
      # Name of volume claim created for Filestore
      claimName: fileserver-claim
      readOnly: false
  restartPolicy: Never
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: myapp
spec:
  host: myapp
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: myapp
spec:
  hosts:
  - myapp.myapp-service.default.svc.cluster.local
  ports:
  - number: 4369
    name: epmd
    protocol: TCP
  - number: 5672
    name: amqp
    protocol: TCP
  - number: 15672
    name: http
    protocol: HTTP
  - number: 25672
    name: inter-node
    protocol: TCP
  - number: 5432
    name: database
    protocol: TCP
---
apiVersion: "authentication.istio.io/v1alpha1"
kind: Policy
metadata:
  name: myapp
spec:
  targets:
  - name: myapp
  peers:
  - mtls: {}`

What am I missing since this does not work?

**Comment by daudn** [comment]
This is my yaml file to run myapp which requires Datastore (port: 5432) and RabbitMQ (port: 5672) 
`
apiVersion: v1
kind: Service
metadata:
  name: myapp-service
  labels:
    app: myapp-service
spec:
  ports:
  - name: rabbitmq
    protocol: TCP
    port: 5672
  - name: datastore
    protocol: TCP
    port: 5432
    targetPort: 5432
  type: NodePort
  selector:
    app: myapp-pod
---
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
spec:
  containers:
    - name: myapp
      image: eu.gcr.io/<projectId>/myapp:latest
      imagePullPolicy: Always
      ports:
        - containerPort: 5672
        - containerPort: 5432
      env:
        - name: MY_NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
      lifecycle:
        preStop:
          exec:
            command: ["python3", "graceful_termination.py"]
      volumeMounts:
      - mountPath: Filestore/
        name: mypvc
  nodeSelector:
    Focus: myapp
  volumes:
  - name: mypvc
    persistentVolumeClaim:
      claimName: fileserver-claim
      readOnly: false
  restartPolicy: Never
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: myapp
spec:
  host: myapp
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: myapp
spec:
  hosts:
  - myapp.myapp-service.default.svc.cluster.local
  ports:
  - number: 4369
    name: epmd
    protocol: TCP
  - number: 5672
    name: amqp
    protocol: TCP
  - number: 15672
    name: http
    protocol: HTTP
  - number: 25672
    name: inter-node
    protocol: TCP
  - number: 5432
    name: database
    protocol: TCP
---
apiVersion: "authentication.istio.io/v1alpha1"
kind: Policy
metadata:
  name: myapp
spec:
  targets:
  - name: myapp
  peers:
  - mtls: {}
`

I get this traceback:

> Traceback (most recent call last):
  File "pbbarcode.py", line 22, in <module>
    key = client.allocate_ids(client.key('analysis_and_barcodes'), 1)[0]
  File "/env/lib/python3.7/site-packages/google/cloud/datastore/client.py", line 509, in allocate_ids
    incomplete_key.project, incomplete_key_pbs
  File "/env/lib/python3.7/site-packages/google/cloud/datastore_v1/gapic/datastore_client.py", line 633, in allocate_ids
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/env/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/env/lib/python3.7/site-packages/google/api_core/retry.py", line 270, in retry_wrapped_func
    on_error=on_error,
  File "/env/lib/python3.7/site-packages/google/api_core/retry.py", line 179, in retry_target
    return target()
  File "/env/lib/python3.7/site-packages/google/api_core/timeout.py", line 214, in func_with_timeout
    return func(*args, **kwargs)
  File "/env/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.ServiceUnavailable: 503 Connect Failed

**Comment by grzesuav** [comment]
Should't port names in `Service` match https://istio.io/docs/setup/kubernetes/additional-setup/requirements/ ?

**Comment by howardjohn** [comment]
We made some bug fixes in this area in 1.3: https://istio.io/news/2019/announcing-1.3/#traffic-management

1.1 is no longer supported: https://istio.io/blog/2019/announcing-1.1-eol/

Please try on recent versions, if you do see the same issues please post all of your config and get config dumps: https://github.com/istio/istio/wiki/Troubleshooting-Istio#collecting-information-2
